docs/specs/pipelines-io.md

```diff
@@ -0,0 +1,196 @@
+
+        remaining -= writable;
+        offset += writable;
+
+        buffer.Advance(writable);
+    }
+}
+```
+
+### Reading
+
+The consumer calls `IPipelineReader.ReadAsync()` which returns with a `ReadableBuffer` when the producer says that data is ready. 
+
+```C#
+private Task Consume(IPipelineReader reader)
+{
+    while (true)
+    {
+        ReadResult result = await reader.ReadAsync();
+        ReadableBuffer buffer = result.Buffer;
+        if (buffer.IsEmpty && result.IsCompleted)
+        {
+            break;
+        }
+
+        // Process the buffer
+        Consume(ref buffer);
+        
+        // Advance the reader past the entire buffer
+        reader.Advance(buffer.End, buffer.End);
+    }
+}
+```
+
+A `ReadableBuffer` is a struct that is a view over one or many `BufferSegments`:
+
+```C#
+public struct ReadCursor
+{
+   public int Index { get; }
+   public BufferSegment Segment { get; }
+}
+
+public struct ReadableBuffer : ISequence<Memory<byte>>
+{
+   public ReadCursor Start { get; }
+   public ReadCursor End { get; }
+   public int Length { get; }
+   
+   public byte Peek();
+   public ReadableBuffer Slice(ReadCursor start, ReadCursor end);
+   public PreservedBuffer Preserve();
+}
+
+public struct PreservedBuffer : IDisposable
+{
+   public ReadableBuffer Buffer { get; }
+}
+```
+
+A `ReadCursor` represents a specific coordinate within a `BufferSegment` and a `ReadableBuffer` represents all of the data between 2 `ReadCursors`, Start and End. The `ReadableBuffer` is a point in time snapshot of the data the producer made available. 
+
+After processing the buffer, consumers call `IPipelineReader.Advance(consumed, observed)` indicate how much of the buffer was consumed and how much of the buffer was observed. The distinction is very important as consumed data can be returned to the system for re-use by another component. Marking observed data lets the consumer control over where it should be notified again (to an extent). 
+
+### Threading
+
+When the producer calls `FlushAsync()` on the `WritableBuffer` that will call the consumer on the producer's thread.
+
+### Buffer ownership
+
+In the examples above, data is pushed from the producer to consumer. The consumer has ownership of the buffer between the call to `IPipelineReader.ReadAsync()` and `IPipelineReader.Advance()`. If the consumers needs ownership of the buffer outside of this scope, an explicit call to `ReadableBuffer.Preserve()` must be made before calling Advance. This will transfer ownership of the buffer to the consumer. `PreservedBuffers` must be disposed.
+
+## Problems
+1. When writing higher level protocol parsers, there is a need for efficient low level routines that can grab a byte at a time, and possible several bytes at a time. Further more, to build successful ecosystem of reusable parsing routines, we'd like these types of parsers to be built on the same primitives. Is `ReadableBuffer` that primitive or do we need something lower? 
```

Comment:
We should talk more, but I think exchange types, as their name implies, are just handles. They need to be good for passing around and representing general concepts. They don't need to be very usable. Stream is our current successful i/o exchange type and it's not very easy to use. We have TextReader building on top of stream to provide usability.
Having said that, I am very much open to improving ISequence. We should talk about what limitations we see today and what scenarios we would like it to support.